### PR TITLE
Fix unserializeArray on nonserialized string

### PR DESCRIPTION
**Pull request by Hanmac**
### Fixed Issues (if relevant)

1. Fixes OpenMage/magento-lts#4383

### Manual testing scenarios (*)

1. Login
2. Put Product into Cart with option Text "abc"
3. Logout
4. Put Product into Cart with option Text "def"
5. Login again, how it tries to Merge the Quotes

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All automated tests passed successfully (all builds are green)
 - [x] Add yourself to contributors list

### CODE REVIEW

**Comment by sreichel** [comment]
There is `Mage_Core_Helper_String::isSerializedArrayOrObject()`

Maybe thats better?

File: app/code/core/Mage/Sales/Model/Quote/Item.php [hunk 1]
```diff
@@ -516,12 +516,16 @@ public function compare($item)
                 // dispose of some options params, that can cramp comparing of arrays
                 if (is_string($itemOptionValue) && is_string($optionValue)) {
                     try {
-                        /** @var Unserialize_Parser $parser */
+                        /**
+                         * @var Mage_Core_Helper_UnserializeArray $parser
+                         * @var Mage_Core_Helper_String $stringHelper
+                         */
```
Comment (kiatng) [inline app/code/core/Mage/Sales/Model/Quote/Item.php:522]: I think it's unnecessary
```suggestion
```

Comment (Hanmac) [reply app/code/core/Mage/Sales/Model/Quote/Item.php:522]: I think IDE would prefer a type hint for the helpers

Comment (kiatng) [reply app/code/core/Mage/Sales/Model/Quote/Item.php:522]: OK, it's harmless. FYI, VSCode + Intelephense has no problem with helpers. I am now using cursor, no issue as well,

**Review by kiatng — APPROVED** [review]

**Comment by sreichel** [comment]
Is it needed? (https://github.com/OpenMage/magento-lts/issues/4383#issuecomment-2502999731)

**Comment by Hanmac** [comment]
> Is it needed? ([#4383 (comment)](https://github.com/OpenMage/magento-lts/issues/4383#issuecomment-2502999731))

Not really?

For this PR, the type is already checked with `if (is_string($itemOptionValue) && is_string($optionValue)) {`
So they can only be string.

This check was for Strings that aren't serialized

**Comment by sreichel** [comment]
If i got your comment right, the isse was fixed with updating your code.

For me everything works, no need for changes. (Something i miss?)

**Comment by Hanmac** [comment]
> If i got your comment right, the isse was fixed with updating your code.
> 
> For me everything works, no need for changes. (Something i miss?)

unserialize still causes an Exception when getting random string
I will make new StackTrace soon

**Comment by Hanmac** [comment]
> If i got your comment right, the isse was fixed with updating your code.
> 
> For me everything works, no need for changes. (Something i miss?)

It relies on the order of the options returned by `Mage_Sales_Model_Quote_Item::getOptions`
and with `info_buyRequest` being first.

If for some reason, I currently can't reproduce, the order gets changed, or `info_buyRequest` gets skipped,
we get a crash like this (Tested with 20.3.0):

```
Exception: Error unserializing data. in /var/www/magento19-native/app/code/core/Mage/Core/Helper/UnserializeArray.php:34
Stack trace:
#0 /var/www/magento19-native/app/code/core/Mage/Sales/Model/Quote/Item.php(523): Mage_Core_Helper_UnserializeArray->unserialize('what')
#1 /var/www/magento19-native/app/code/core/Mage/Sales/Model/Quote.php(1876): Mage_Sales_Model_Quote_Item->compare(Object(Mage_Sales_Model_Quote_Item))
#2 /var/www/magento19-native/app/code/core/Mage/Checkout/Model/Session.php(294): Mage_Sales_Model_Quote->merge(Object(Mage_Sales_Model_Quote))
#3 /var/www/magento19-native/app/code/core/Mage/Checkout/Model/Observer.php(32): Mage_Checkout_Model_Session->loadCustomerQuote()
#4 /var/www/magento19-native/app/code/core/Mage/Core/Model/App.php(1448): Mage_Checkout_Model_Observer->loadCustomerQuote(Object(Varien_Event_Observer))
#5 /var/www/magento19-native/app/code/core/Mage/Core/Model/App.php(1426): Mage_Core_Model_App->_callObserverMethod(Object(Mage_Checkout_Model_Observer), 'loadCustomerQuo...', Object(Varien_Event_Observer), 'loadCustomerQuo...')
#6 /var/www/magento19-native/app/Mage.php(509): Mage_Core_Model_App->dispatchEvent('customer_login', Array)
#7 /var/www/magento19-native/app/code/core/Mage/Customer/Model/Session.php(250): Mage::dispatchEvent('customer_login', Array)
#8 /var/www/magento19-native/app/code/core/Mage/Customer/Model/Session.php(235): Mage_Customer_Model_Session->setCustomerAsLoggedIn(Object(Mage_Customer_Model_Customer))
#9 /var/www/magento19-native/app/code/core/Mage/Customer/controllers/AccountController.php(151): Mage_Customer_Model_Session->login('tester@testshop...', 'Test123')
#10 /var/www/magento19-native/app/code/core/Mage/Core/Controller/Varien/Action.php(421): Mage_Customer_AccountController->loginPostAction()
#11 /var/www/magento19-native/app/code/core/Mage/Core/Controller/Varien/Router/Standard.php(255): Mage_Core_Controller_Varien_Action->dispatch('loginPost')
#12 /var/www/magento19-native/app/code/core/Mage/Core/Controller/Varien/Front.php(181): Mage_Core_Controller_Varien_Router_Standard->match(Object(Mage_Core_Controller_Request_Http))
#13 /var/www/magento19-native/app/code/core/Mage/Core/Model/App.php(358): Mage_Core_Controller_Varien_Front->dispatch()
#14 /var/www/magento19-native/app/Mage.php(743): Mage_Core_Model_App->run(Array)
#15 /var/www/magento19-native/index.php(56): Mage::run('', 'store')
#16 {main}
```

**Review by sreichel — APPROVED** [review]

**Pull request merged**